src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,299 @@
+{
+    mount_config->format_if_mount_failed = true;
+    mount_config->max_files = 8;
+    mount_config->allocation_unit_size = 512;
+    // TODO: make it configurable: disk_status_check_enable = false
+}
+
+static term nif_esp_mount(Context *ctx, int argc, term argv[])
+{
+    term source_term = argv[0];
+    term target_term = argv[1];
+    term filesystem_type_term = argv[2];
+    term opts_term = argv[3];
+
+    int str_ok;
+    char *source = interop_term_to_string(source_term, &str_ok);
+    if (!str_ok) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    char *target = interop_term_to_string(target_term, &str_ok);
+    if (!str_ok || strlen(target) > 8) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    term fat_term
+        = globalcontext_existing_term_from_atom_string(ctx->global, ATOM_STR("\x3", "fat"));
+    if (term_is_invalid_term(fat_term) || filesystem_type_term != fat_term) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    if (!term_is_list(opts_term) && !term_is_map(opts_term)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    struct ESP32PlatformData *platform = ctx->global->platform_data;
+
+    esp_vfs_fat_mount_config_t mount_config = {};
+    opts_to_fatfs_mount_config(opts_term, &mount_config);
+
+    esp_err_t ret = -1;
+
+    const char *part_by_name_prefix = "/dev/partition/by-name/";
+    int part_by_name_len = strlen(part_by_name_prefix);
+    if (!strncmp(part_by_name_prefix, source, part_by_name_len)) {
+        mount_config.allocation_unit_size = CONFIG_WL_SECTOR_SIZE;
+
+        struct MountedFS *mount = malloc(sizeof(struct MountedFS));
+        mount->misc_entry.entry_type = (uintptr_t) nif_esp_mount;
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSPIFlash;
+        synclist_append(&platform->misc_entries, &mount->misc_entry.list_head);
+
+#if ESP_IDF_VERSION_MAJOR >= 5
+        ret = esp_vfs_fat_spiflash_mount_rw_wl(
+            target, source + part_by_name_len, &mount_config, &mount->handle.wl);
+#else
+        ret = esp_vfs_fat_spiflash_mount(
+            target, source + part_by_name_len, &mount_config, &mount->handle.wl);
+#endif
+
+// C3 doesn't support this
+#ifdef SDMMC_SLOT_CONFIG_DEFAULT
+    } else if (!strcmp(source, "sdmmc")) {
+        mount_config.allocation_unit_size = 512;
+
+        sdmmc_host_t host_config = SDMMC_HOST_DEFAULT();
+        sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
+
+        struct MountedFS *mount = malloc(sizeof(struct MountedFS));
+        mount->misc_entry.entry_type = (uintptr_t) nif_esp_mount;
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSDMMC;
+        synclist_append(&platform->misc_entries, &mount->misc_entry.list_head);
+
+        ret = esp_vfs_fat_sdmmc_mount(
+            target, &host_config, &slot_config, &mount_config, &mount->handle.card);
+#endif
+
+    } else if (!strcmp(source, "sdspi")) {
```

Comment:
Help for manual testing this might be appreciated.

Reply:
I have tried with several devices but it seems none of the ones I own use SDMMC_HOST_DEFAULT() or SPI_HOST_DEFAULT() pins, so for wide spread compatibility these will need to be configurable.